src/wallet/wallet2.cpp

```diff
@@ -4603,71 +4603,89 @@ std::vector<wallet2::pending_tx> wallet2::create_transactions_2(std::vector<cryp
-    }
-  }
   for (uint32_t i : subaddr_indices)
-    LOG_PRINT_L2("Spending from subaddress index " << i);
+    LOG_PRINT_L2("Candidate subaddress index for spending: " << i);
 
-  // gather all dust and non-dust outputs of specified subaddress
+  // gather all dust and non-dust outputs belonging to specified subaddresses
+  size_t num_nondust_outputs = 0;
+  size_t num_dust_outputs = 0;
   for (size_t i = 0; i < m_transfers.size(); ++i)
   {
     const transfer_details& td = m_transfers[i];
     if (!td.m_spent && (use_rct ? true : !td.is_rct()) && is_transfer_unlocked(td) && td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
     {
+      const uint32_t index_minor = td.m_subaddr_index.minor;
+      auto find_predicate = [&index_minor](const std::pair<uint32_t, std::vector<size_t>>& x) { return x.first == index_minor; };
       if ((td.is_rct()) || is_valid_decomposed_amount(td.amount()))
-        unused_transfers_indices.push_back(i);
+      {
+        auto found = std::find_if(unused_transfers_indices_per_subaddr.begin(), unused_transfers_indices_per_subaddr.end(), find_predicate);
+        if (found == unused_transfers_indices_per_subaddr.end())
+        {
+          unused_transfers_indices_per_subaddr.push_back({index_minor, {i}});
+        }
+        else
+        {
+          found->second.push_back(i);
+        }
+        ++num_nondust_outputs;
+      }
       else
-        unused_dust_indices.push_back(i);
+      {
+        auto found = std::find_if(unused_dust_indices_per_subaddr.begin(), unused_dust_indices_per_subaddr.end(), find_predicate);
+        if (found == unused_dust_indices_per_subaddr.end())
+        {
+          unused_dust_indices_per_subaddr.push_back({index_minor, {i}});
+        }
+        else
+        {
+          found->second.push_back(i);
+        }
+        ++num_dust_outputs;
+      }
     }
   }
 
-  LOG_PRINT_L2("Starting with " << unused_transfers_indices.size() << " non-dust outputs and " << unused_dust_indices.size() << " dust outputs");
+  // shuffle & sort output indices
+  {
+    std::random_device rd;
+    std::mt19937 g(rd());
+    std::shuffle(unused_transfers_indices_per_subaddr.begin(), unused_transfers_indices_per_subaddr.end(), g);
+    std::shuffle(unused_dust_indices_per_subaddr.begin(), unused_dust_indices_per_subaddr.end(), g);
+    auto sort_predicate = [this] (const std::pair<uint32_t, std::vector<size_t>>& x, const std::pair<uint32_t, std::vector<size_t>>& y)
+    {
+      uint64_t sum_x = 0;
+      for (size_t i : x.second)
+        sum_x += m_transfers[i].amount();
+      uint64_t sum_y = 0;
+      for (size_t i : y.second)
+        sum_y += m_transfers[i].amount();
+      return sum_x > sum_y;
+    };
+    std::sort(unused_transfers_indices_per_subaddr.begin(), unused_transfers_indices_per_subaddr.end(), sort_predicate);
+    std::sort(unused_dust_indices_per_subaddr.begin(), unused_dust_indices_per_subaddr.end(), sort_predicate);
+  }
```

Comment:
The shuffling prior to the sort is for randomly choosing a subaddress out of a set of subaddresses that have the same unlocked balance (note that the randomized ordering is preserved after the sort). For example, if the addresses 1 & 2 have unlocked balance of 2.0 XMR while the address 3 has unlocked balance of 1.0 XMR, the algorithm chooses one of the following orders of addresses at random: 1->2->3 or 2->1->3.

Yes, the sort predicate computes the sum of the selected transfers for the given subaddress index. Maybe I should modify the code so that this redundant computation is avoided.